tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+your entire code base.
+
+## Builder Internals
+
+In the examples you've seen so far, we've only ever looked at stages,
+expressions, and of course the `Pipeline` objects. But how do we end up with an
+aggregation pipeline that the server understands? The objects created with the
+various factory methods are value holder objects. They don't contain any logic,
+but they allow us to know what types an expression will resolve to.
+
+For an example, let's take the `$hour` operator. When you use
+`Expression::hour`, you will receive an instance of an `HourOperator` class.
+This class implements an `OperatorInterface`, telling the builder that this is
+an operator that can be used in aggregation pipeline. It also implements a
+`ResolvesToInt` interface, as we always know that evaluating the expression
+results in an integer value. The required `date` parameter of the operator is a
+date, which in the builder is one of many things. It could be a BSON
+`UTCDateTime` instance, but it could also be the result of any operator that
+returns a date, e.g. `$dateFromString`.
+
+Now that we know about these value holder objects, we still need to make sure
+the server knows what we're talking about. When you call `Collection::aggregate`
+with the pipeline you built, what happens internally to it? Here, a series of
+encoders springs into action. We use a single entry point, the `BuilderEncoder`
+class. This class contains multiple encoders that are able to handle all
+pipeline stages, operators, and accumulators and transform them into their BSON
+representations.
+
+This allows us to keep the logic customizable. For example, the Doctrine MongoDB
+ODM allows users to specify different names for fields in the database. In turn,
+it needs to convert the name of a property in the mapped class to the name of
+the field in the database. Such a feature could easily be built by creating a
+custom encoder for all `fieldPath` expressions, and changing the field path
+accordingly.
+
+When creating a `MongoDB\Client` instance, you can now pass an additional
+`builderEncoder` option in the `$driverOptions` argument. This specifies the
+encoder used to encode aggregation pipelines, but also query objects. All
+`Database` and `Client` instances inherit this value from the client, but you
+can override it through the options when fetching such an instance. This allows
+you to have your custom logic applied whenever pipelines or queries are encoded
+for the server.
+
+With factories, value holders, and encoders, we wanted to ensure that creating
+the builder does not turn into a repetitive chore. As you can imagine, many
+operators will mostly consist of the same logic, resulting in tons of code
+duplication. To make matters worse, every new server version adds some new
+operators or even stages, so we wanted to make sure that we can easily expand
+the builder.
+
+We could try to rely on generative AI to help us with this, but this only goes
```

Comment:
Please tell me "generative AI" here was intended as an SEO play for the blog post. Did we, at any point, actually consider using it for this project? IIRC, it was always defining the aggregation syntax in YAML and employing code generation.

Reply:
Yes and no. I may or may not have used GitHub CoPilot to protect my sanity in [one](https://github.com/doctrine/mongodb-odm/pull/2514) [or](https://github.com/doctrine/mongodb-odm/pull/2513) [more](https://github.com/doctrine/mongodb-odm/pull/2517) larger pull requests introducing aggregation builder functionality in the Doctrine ODM recently.

SEO is a nice benefit that I'll happily take.